src/lib/verifyAppleIdToken.ts

```diff
@@ -1,45 +1,34 @@
-import * as jwt from "jsonwebtoken";
-import * as jwksClient from "jwks-rsa";
-import { VerifyAppleIdTokenParams, VerifyAppleIdTokenResponse } from "./types";
+import { createRemoteJWKSet, decodeProtectedHeader, jwtVerify } from "jose";
+import { VerifyAppleIdTokenParams } from "./types";
 
 export const APPLE_BASE_URL = "https://appleid.apple.com";
 export const JWKS_APPLE_URI = "/auth/keys";
 
-export const getApplePublicKey = async (kid: string) => {
-  const client = jwksClient({
-    cache: true,
-    jwksUri: `${APPLE_BASE_URL}${JWKS_APPLE_URI}`,
+export const getApplePublicKey = async (kid: string, alg: string) => {
+  const JWKS = createRemoteJWKSet(new URL(`${APPLE_BASE_URL}${JWKS_APPLE_URI}`));
+  const key = await JWKS({
+    alg,
+    kid,
   });
-  const key = await new Promise<jwksClient.SigningKey>((resolve, reject) => {
-    client.getSigningKey(kid, (error, result) => {
-      if (error) {
-        return reject(error);
-      }
-      return resolve(result);
-    });
-  });
-  return key.getPublicKey();
+  return key;
 };
 
 export const verifyToken = async (params: VerifyAppleIdTokenParams) => {
-  const decoded = jwt.decode(params.idToken, { complete: true });
-  const { kid, alg } = decoded.header;
+  const { alg, kid } = decodeProtectedHeader(params.idToken);
 
-  const applePublicKey = await getApplePublicKey(kid);
-  const jwtClaims = jwt.verify(params.idToken, applePublicKey, {
-    algorithms: [alg as jwt.Algorithm],
-    nonce: params.nonce,
-  }) as VerifyAppleIdTokenResponse;
+  const applePublicKey = await getApplePublicKey(kid, alg);
+
+  const { payload: jwtClaims } = await jwtVerify(params.idToken, applePublicKey);
+
+  if (jwtClaims?.nonce !== params.nonce) {
```

Comment:
I am not sure about the response of the `jwtClaims?.nonce` - is it `null` or `undefined` when it is empty? Because if it is `null` the condition will fail (`params.nonce` is `undefined` or `string`). I suggest to use optional chaining to be 100 % sure about it.

```suggestion
  if ((jwtClaims?.nonce ?? undefined) !== params.nonce) {
```